milli/src/search/mod.rs

```diff
@@ -284,20 +285,60 @@ pub fn word_derivations<'c>(
         Entry::Occupied(entry) => Ok(entry.into_mut()),
         Entry::Vacant(entry) => {
             let mut derived_words = Vec::new();
-            let dfa = build_dfa(word, max_typo, is_prefix);
-            let mut stream = fst.search_with_state(&dfa).into_stream();
+            if max_typo == 0 {
+                if is_prefix {
+                    let prefix = Str::new(word).starts_with();
+                    let mut stream = fst.search(prefix).into_stream();
+
+                    while let Some(word) = stream.next() {
+                        let word = std::str::from_utf8(word)?;
+                        derived_words.push((word.to_string(), 0));
+                    }
+                } else {
+                    let automaton = Str::new(word);
+                    let mut stream = fst.search(automaton).into_stream();
+                    while let Some(word) = stream.next() {
+                        let word = std::str::from_utf8(word)?;
+                        derived_words.push((word.to_string(), 0));
+                    }
+                }
+            } else {
+                if max_typo == 1 {
+                    let dfa = build_dfa(word, 1, is_prefix);
+                    let starts = Str::new(get_first(word)).starts_with();
+                    let mut stream = fst.search_with_state(starts.intersection(&dfa)).into_stream();
+
+                    while let Some((word, state)) = stream.next() {
+                        let word = std::str::from_utf8(word)?;
+                        let distance = dfa.distance(state.1);
+                        derived_words.push((word.to_string(), distance.to_u8()));
+                    }
+                } else {
+                    let starts = Str::new(get_first(word)).starts_with();
+                    let first = build_dfa(word, 1, is_prefix).intersection((&starts).complement());
+                    let second = build_dfa(word, 2, is_prefix).intersection(&starts);
+                    let automaton = first.union(second);
 
-            while let Some((word, state)) = stream.next() {
-                let word = std::str::from_utf8(word)?;
-                let distance = dfa.distance(state);
-                derived_words.push((word.to_string(), distance.to_u8()));
-            }
+                    let mut stream = fst.search(automaton).into_stream();
 
+                    while let Some(word) = stream.next() {
+                        let word = std::str::from_utf8(word)?;
+                        derived_words.push((word.to_string(), 2));
+                    }
+                }
+            }
             Ok(entry.insert(derived_words))
         }
     }
 }
 
+fn get_first(s: &str) -> &str {
+    match s.chars().next() {
+        Some(c) => &s[..c.len_utf8()],
+        None => s,
+    }
+}
```

Comment:
Do you want to return an empty string when you are unable to return the first char as a string?

Reply:
I suppose the beginning of an empty string is the strong itself? Seemed logical no me, what would you rather do?

Reply:
This function does not return the beginning of a string but rather the first char of it, as it is empty, we should either return None, panics internally or maybe return an error that we can easily propagate with `?`. I would go for a panic personally, as it is not possible to have an empty string at this point.

Reply:
panic it is then